README.md
kpp-interfacestep
=================

This package implements a graphical user interface for the
hpp-wholebody-step-planner package.


NEWS
                                                                -*- outline -*-

New in 0.3, 2013-10-08:
* Add hpp-util dependency.
* Fix compilation warnings.
* Synchronize cmake submodule.

New in 0.2, 2013-09-26:
* Update to kpp-interface 2.0
* Update to hpp-wholebody-step-planner 0.2

New in 0.1, kpp-interfacestep:
* First release.


tests/simple-test.cc
// Copyright (C) 2009 by Sebastien Dalibard.
//
// This file is part of the kpp-interfacestep.
//
// kpp-interfacestep is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// kpp-interfacestep is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with kpp-interfacestep.  If not, see <http://www.gnu.org/licenses/>.

#include "common.hh"

int run_test ();

int run_test ()
{
  //FIXME: make a real test.
  return 0;
}

GENERATE_TEST ()


src/command-grabobject.cc
#include <vector>

#include "KineoModel/kppRotationJointComponent.h"
#include "KineoController/kppSetPropertyCommand.h"
#include "KineoController/kppDocument.h"
#include "KineoModel/kppComponentParameter.h"
#include "KineoModel/kppNumberParameter.h"
#include "KineoModel/kppValue.h"
#include "KineoModel/kppComponentClassFilter.h"
#include "KineoModel/kppComponent.h"
#include "KineoModel/kppSolidComponent.h"
#include "KineoModel/kppPathNode.h"
#include "KineoModel/kppPathComponent.h"
#include "KineoModel/kppModelTree.h"
#include "KineoGUI/kppMainWindowController.h" 
#include "KineoController/kppDocument.h"
#include "KineoGUI/kppMainWindowUICommandFactory.h"

# include <hpp/util/debug.hh>
# include <hpp/wholebody-step-planner/planner.hh>

#include "kpp/interfacestep/command-grabobject.hh"
#include "kpp/interfacestep/interface.hh"

namespace kpp
{
  namespace interfaceStep
  {
    // ==========================================================================

    CommandGrabObject::CommandGrabObject(Interface *kpp)
    {

      attKpp = kpp;

    }


    // ==========================================================================

    CommandGrabObject::CommandGrabObject(const CommandGrabObject& inCommand) :
      CkppCommand(inCommand)
    {
      attKpp = inCommand.attKpp;

    }


    // ==========================================================================

    CommandGrabObject::~CommandGrabObject()
    {
      weakPtr_.reset();
    }


    // ==========================================================================

    CommandGrabObjectShPtr CommandGrabObject::create(Interface *kpp)
    {
      CommandGrabObject*  ptr = new CommandGrabObject(kpp);
      CommandGrabObjectShPtr shPtr(ptr);
  
      if(KD_OK != ptr->init(shPtr))
	{
	  shPtr.reset();
	}

      return shPtr;
    }

    // ==========================================================================

    CommandGrabObjectShPtr CommandGrabObject::createCopy(const CommandGrabObjectConstShPtr& inCommand)
    {
      CommandGrabObject*  ptr = new CommandGrabObject(*inCommand);
      CommandGrabObjectShPtr shPtr(ptr);
  
      if(KD_OK != ptr->init(shPtr))
	{
	  shPtr.reset();
	}
  
      return shPtr;
    }


    // ==========================================================================

    ktStatus CommandGrabObject::init(const CommandGrabObjectWkPtr& inWeakPtr)
    {
      ktStatus success = CkppCommand::init(inWeakPtr);
  
      if(KD_OK == success)
	{
	  weakPtr_ = inWeakPtr;
	}


      return success;
    }


    // ==========================================================================


    CkppCommandShPtr CommandGrabObject::clone() const
    {
      return CommandGrabObject::createCopy(weakPtr_.lock());
    }


    // ==========================================================================

    bool CommandGrabObject::isUndoable() const
    {
      return true;
    }


    // ==========================================================================

    unsigned int CommandGrabObject::countParameters() const
    {
      return PARAMETER_COUNT;
    }


    // ==========================================================================

    CkppParameterConstShPtr CommandGrabObject::parameter(unsigned int inRank) const
    {
 
      CkppParameterShPtr result;
      CkppComponentShPtr nullComponent;

      switch(inRank) {   
      default:
      case SOLID_COMPONENT : result = CkppComponentParameter::create("Solid Component", CkppComponentClassFilter<CkppSolidComponent>());    break;
	KIT_ASSERT( false );
      }
  
      return result;
    }


    // ==========================================================================

    ktStatus CommandGrabObject::doExecute()
    {
      CkppComponentShPtr component(paramValue(parameter(SOLID_COMPONENT)).componentValue());

      CkppSolidComponentShPtr solidComponent = KIT_DYNAMIC_PTR_CAST(CkppSolidComponent, component);

      using namespace hpp::wholeBodyStepPlanner;
	
      Planner *planner = dynamic_cast<Planner*> ( attKpp->hppPlanner() );

      CkitMat4 absPos;
      solidComponent->getAbsolutePosition ( absPos );
      vectorN target ( 3 );
      target[0] = absPos ( 0,3 );
      target[1] = absPos ( 1,3 );
      target[2] = absPos ( 2,3 );

      planner->generateGoalConfig (target[0], target[1], target[2], 1);

      return KD_OK;
    }

    // ==========================================================================
  }
}


src/command-dynamicpath.cc
# include <KineoModel/kppPathComponent.h>
# include <KineoModel/kppComponentClassFilter.h>
# include <KineoModel/kppComponentParameter.h>

# include <hpp/wholebody-step-planner/planner.hh>

# include <kpp/interfacestep/command-dynamicpath.hh>
# include <kpp/interfacestep/interface.hh>


namespace kpp
{
  namespace interfaceStep
  {
    CommandDynamicPath::CommandDynamicPath ( Interface * kpp)
    {
      kpp_ = kpp;
    }

    CommandDynamicPath::CommandDynamicPath(const CommandDynamicPath& inCommand):
      CkppCommand ( inCommand )
    {
      kpp_ = inCommand.kpp_;
    }

    CommandDynamicPath::~CommandDynamicPath()
    {
      wkPtr_.reset();
    }

    CommandDynamicPathShPtr 
    CommandDynamicPath::create(Interface *kpp)
    {
      CommandDynamicPath * ptr = new CommandDynamicPath (kpp);
      CommandDynamicPathShPtr shPtr (ptr);
      if (ptr->init (shPtr) != KD_OK)
	shPtr.reset();
      return shPtr;
    }

    CommandDynamicPathShPtr
    CommandDynamicPath::createCopy( const  CommandDynamicPathConstShPtr& inCommand)
    {
      CommandDynamicPath * ptr = new CommandDynamicPath( *inCommand);
      CommandDynamicPathShPtr shPtr (ptr);
      if (ptr->init (shPtr) != KD_OK)
	shPtr.reset();
      return shPtr;
    }

    ktStatus
    CommandDynamicPath::init( const CommandDynamicPathWkPtr &inWkPtr)
    {
      ktStatus res = CkppCommand::init (inWkPtr);
      if ( res == KD_OK)
	wkPtr_ = inWkPtr;
      return res;
    }

    CkppCommandShPtr
    CommandDynamicPath::clone() const
    {
      return CommandDynamicPath::createCopy ( wkPtr_.lock() );
    }

    bool 
    CommandDynamicPath::isUndoable() const
    {
      return true;
    }

    unsigned int
    CommandDynamicPath::countParameters() const
    {
      return PARAMETER_COUNT;
    }

    CkppParameterConstShPtr
    CommandDynamicPath::parameter ( unsigned int inRank ) const
    {
      CkppParameterShPtr result;
      CkppComponentShPtr nullComponent;
      switch ( inRank )
	{
	default:
	case PATH_COMPONENT : result = CkppComponentParameter::create ( "Path Component", CkppComponentClassFilter<CkppPathComponent>() );    break;
	  KIT_ASSERT ( false );
	}
      return result;
    }

    ktStatus
    CommandDynamicPath::doExecute()
    {
      using namespace hpp::wholeBodyStepPlanner;

      CkppComponentShPtr component ( paramValue ( parameter ( PATH_COMPONENT ) ).componentValue() ); 
      
      CkppPathComponentShPtr pathComponent = 
	KIT_DYNAMIC_PTR_CAST ( CkppPathComponent, component );

      Planner * planner = 
	dynamic_cast<Planner*> (kpp_->hppPlanner());
      
      if (planner)
	{
	  CkwsPathShPtr dynamicPath = planner-> findDynamicPath(pathComponent->kwsPath());
	  if (dynamicPath)
	    planner->hppProblem(0)->addPath(dynamicPath);
	}
      else
	return KD_ERROR;

      return KD_OK;
    }
  }
}


src/command-setinitgoal.cc
# include <KineoModel/kppPathComponent.h>
# include <KineoModel/kppComponentClassFilter.h>
# include <KineoModel/kppComponentParameter.h>

# include <hpp/wholebody-step-planner/planner.hh>

# include <kpp/interfacestep/command-setinitgoal.hh>
# include <kpp/interfacestep/interface.hh>


namespace kpp
{
  namespace interfaceStep
  {
    CommandSetInitGoal::CommandSetInitGoal ( Interface * kpp)
    {
      kpp_ = kpp;
    }

    CommandSetInitGoal::CommandSetInitGoal(const CommandSetInitGoal& inCommand):
      CkppCommand ( inCommand )
    {
      kpp_ = inCommand.kpp_;
    }

    CommandSetInitGoal::~CommandSetInitGoal()
    {
      wkPtr_.reset();
    }

    CommandSetInitGoalShPtr 
    CommandSetInitGoal::create(Interface *kpp)
    {
      CommandSetInitGoal * ptr = new CommandSetInitGoal (kpp);
      CommandSetInitGoalShPtr shPtr (ptr);
      if (ptr->init (shPtr) != KD_OK)
	shPtr.reset();
      return shPtr;
    }

    CommandSetInitGoalShPtr
    CommandSetInitGoal::createCopy( const  CommandSetInitGoalConstShPtr& inCommand)
    {
      CommandSetInitGoal * ptr = new CommandSetInitGoal( *inCommand);
      CommandSetInitGoalShPtr shPtr (ptr);
      if (ptr->init (shPtr) != KD_OK)
	shPtr.reset();
      return shPtr;
    }

    ktStatus
    CommandSetInitGoal::init( const CommandSetInitGoalWkPtr &inWkPtr)
    {
      ktStatus res = CkppCommand::init (inWkPtr);
      if ( res == KD_OK)
	wkPtr_ = inWkPtr;
      return res;
    }

    CkppCommandShPtr
    CommandSetInitGoal::clone() const
    {
      return CommandSetInitGoal::createCopy ( wkPtr_.lock() );
    }

    bool 
    CommandSetInitGoal::isUndoable() const
    {
      return true;
    }

    unsigned int
    CommandSetInitGoal::countParameters() const
    {
      return PARAMETER_COUNT;
    }

    CkppParameterConstShPtr
    CommandSetInitGoal::parameter ( unsigned int inRank ) const
    {
      CkppParameterShPtr result;
      CkppComponentShPtr nullComponent;
      switch ( inRank )
	{
	default:
	case PATH_COMPONENT : result = CkppComponentParameter::create ( "Path Component", CkppComponentClassFilter<CkppPathComponent>() );    break;
	  KIT_ASSERT ( false );
	}
      return result;
    }

    ktStatus
    CommandSetInitGoal::doExecute()
    {
      using namespace hpp::wholeBodyStepPlanner;

      CkppComponentShPtr component ( paramValue ( parameter ( PATH_COMPONENT ) ).componentValue() ); 
      
      CkppPathComponentShPtr pathComponent = 
	KIT_DYNAMIC_PTR_CAST ( CkppPathComponent, component );

      Planner * planner = 
	dynamic_cast<Planner*> (kpp_->hppPlanner());
      
      if (planner)
	planner->initAndGoalConfig (pathComponent->kwsPath());
      else
	return KD_ERROR;

      return KD_OK;
    }
  }
}


src/command-removeconstraints.cc
# include <KineoModel/kppPathComponent.h>
# include <KineoModel/kppComponentClassFilter.h>
# include <KineoModel/kppComponentParameter.h>

# include <hpp/constrained/kws-constraint.hh>
# include <hpp/wholebody-step-planner/planner.hh>

# include "kpp/interfacestep/command-removeconstraints.hh"
# include "kpp/interfacestep/interface.hh"


namespace kpp
{
  namespace interfaceStep
  {
    CommandRemoveConstraints::CommandRemoveConstraints ( Interface * kpp)
    {
      kpp_ = kpp;
    }

    CommandRemoveConstraints::CommandRemoveConstraints(const CommandRemoveConstraints& inCommand):
      CkppCommand ( inCommand )
    {
      kpp_ = inCommand.kpp_;
    }

    CommandRemoveConstraints::~CommandRemoveConstraints()
    {
      wkPtr_.reset();
    }

    CommandRemoveConstraintsShPtr 
    CommandRemoveConstraints::create(Interface *kpp)
    {
      CommandRemoveConstraints * ptr = new CommandRemoveConstraints (kpp);
      CommandRemoveConstraintsShPtr shPtr (ptr);
      if (ptr->init (shPtr) != KD_OK)
	shPtr.reset();
      return shPtr;
    }

    CommandRemoveConstraintsShPtr
    CommandRemoveConstraints::createCopy( const  CommandRemoveConstraintsConstShPtr& inCommand)
    {
      CommandRemoveConstraints * ptr = new CommandRemoveConstraints( *inCommand);
      CommandRemoveConstraintsShPtr shPtr (ptr);
      if (ptr->init (shPtr) != KD_OK)
	shPtr.reset();
      return shPtr;
    }

    ktStatus
    CommandRemoveConstraints::init( const CommandRemoveConstraintsWkPtr &inWkPtr)
    {
      ktStatus res = CkppCommand::init (inWkPtr);
      if ( res == KD_OK)
	wkPtr_ = inWkPtr;
      return res;
    }

    CkppCommandShPtr
    CommandRemoveConstraints::clone() const
    {
      return CommandRemoveConstraints::createCopy ( wkPtr_.lock() );
    }

    bool 
    CommandRemoveConstraints::isUndoable() const
    {
      return true;
    }

    unsigned int
    CommandRemoveConstraints::countParameters() const
    {
      return PARAMETER_COUNT;
    }

    CkppParameterConstShPtr
    CommandRemoveConstraints::parameter ( unsigned int inRank ) const
    {
      CkppParameterShPtr result;
      CkppComponentShPtr nullComponent;
      switch ( inRank )
	{
	default:
	case PATH_COMPONENT : result = CkppComponentParameter::create ( "Path Component", CkppComponentClassFilter<CkppPathComponent>() );    break;
	  KIT_ASSERT ( false );
	}
      return result;
    }

    ktStatus
    CommandRemoveConstraints::doExecute()
    {
      using namespace hpp::wholeBodyStepPlanner;

      Planner * planner = 
	dynamic_cast<Planner*> (kpp_->hppPlanner());
      
      if (planner)
	  planner->humanoidRobot ()->userConstraints ()->remove (planner->wholeBodyConstraint ());
      else
	return KD_ERROR;

      return KD_OK;
    }
  }
}


src/interface.cc
# include <fstream>

# include <KineoGUI/kppMainWindowUICommandFactory.h>
# include <KineoController/kppUICommandList.h>
# include <KineoController/kppUICommand.h>

# include <jrl/mal/matrixabstractlayer.hh>

# include <hpp/util/debug.hh>
# include <hpp/wholebody-step-planner/planner.hh>
# include <hpp/corbaserver/wholebody-step/server.hh>

# include <kpp/interfacestep/interface.hh>
# include <kpp/interfacestep/command-grabobject.hh>
# include <kpp/interfacestep/command-setinitgoal.hh>
# include <kpp/interfacestep/command-dynamicpath.hh>
# include <kpp/interfacestep/command-removeconstraints.hh>

# include "config.h"

namespace kpp
{
  namespace interfaceStep
  {
    Interface::Interface (Planner * planner)
      : CkppInterface (planner),
	wholeBodyStepPlanner_ (planner)
    {
      char** argv = (char**) malloc (sizeof (char*));
      argv[0] = strdup ("Kite");
      server_ = new Server (1, argv, false, "child");
      server_->setPlanner (wholeBodyStepPlanner_);
      try {
	startCorbaServer ();
	hppDout (info, "Successfully started hpp-corbaserver.");
      } catch (const std::exception& exc) {
	hppDout (error, "Failed to start hpp-corbaserver");
      }
      try {
	server_->startCorbaServer("hpp", "plannerContext",
				  "hpp", "wholeBodyStep");
	hppDout (info, "Successfully started wholebody-step corba server.");
      } catch (const std::exception& exc) {
	hppDout (error, "Failed to start wholebody-step corba server.");
      }
    }

    Interface::~Interface ()
    {}

    InterfaceShPtr
    Interface::create ()
    {
      double samplingPeriod = 0.05;
      Planner * planner = new Planner(samplingPeriod);

      KIT_ASSERT(planner);
      planner->setFootPrintLimits(-0.2,0.2,-0.25,-0.13,-M_PI /4,0.1);

      InterfaceShPtr ptr (new Interface (planner));
      InterfaceWkPtr wkPtr (ptr);

      if (ptr->init (wkPtr) != KD_OK)
	ptr.reset ();
      return ptr;
    }

    ktStatus
    Interface::init (const InterfaceWkPtr& weakPtr)
    {
      if (CkppInterface::init (weakPtr) != KD_OK)
	return KD_ERROR;
      weakPtr_ = weakPtr;
      return KD_OK;
    }

    std::string
    Interface::name () const
    {
      return PACKAGE_STRING;
    }

    void
    Interface::getMenuUICommandLists(const CkppMainWindowUICommandFactoryConstShPtr & inCommandFactory,
				     std::vector< CkppUICommandListShPtr > & outMenuCommandListVector)
    {
      CkppInterface::getMenuUICommandLists(inCommandFactory, outMenuCommandListVector);

      CkppUICommandListShPtr hppUICommandList = CkppUICommandList::create ( "Step Planner" );

      commandGrabObject_ = CkppUICommand::create ( CommandGrabObject::create ( this ),
						    inCommandFactory->environment(),
						    "Generate goal configs",
						    "Generate goal configs");

      commandSetInitGoal_ = CkppUICommand::create ( CommandSetInitGoal::create ( this ),
						    inCommandFactory->environment(),
						    "Set Init/Goal configs",
						    "Set Init/Goal configs");
      commandRemoveConstraints_ =  CkppUICommand::create ( CommandRemoveConstraints::create ( this ),
						     inCommandFactory->environment(),
						     "Remove Foot Constraints",
						     "Remove Foot Constraints");
      commandDynamicPath_ =  CkppUICommand::create ( CommandDynamicPath::create ( this ),
						     inCommandFactory->environment(),
						     "Animate Path",
						     "Animate Path");

      hppUICommandList->appendCommand(commandGrabObject_);
      hppUICommandList->appendCommand(commandSetInitGoal_);
      hppUICommandList->appendCommand(commandRemoveConstraints_);
      hppUICommandList->appendCommand(commandDynamicPath_);

      outMenuCommandListVector.push_back ( hppUICommandList );
    }


  }
}

extern "C"
int
initializeModule (CkppModuleInterfaceShPtr& o_moduleInterface)
{
  using namespace kpp::interfaceStep;

  o_moduleInterface = Interface::create ();

  if (o_moduleInterface)
    return 0;

  //FIXME: for some reason, a segv happens if the module initialization
  // fails. As the user is doomed anyway, let's exit with a nice error
  // message.

  return 1;
}
